blackbaud-community/stache-cli
# Build & Deploy with Travis-CI

## Purpose:

To create a workflow for continuous integration deployments with Travis-CI. There should be marginal effort required from our basic--and advanced--Stache users to deploy their sites in a way that is consistent and adheres to best practices. This particular workflow borrows heavily from [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow), and _enforces pull requests as the only method for deployment_.
## Overview:

Azure sites "watch" a Github branch and trigger a deployment when that branch receives a push notification. Currently, Azure will _build and deploy_ the site; this feature branch hopes to remove the build step from Azure to speed up the deployment process and catch any issues that occur during the process. Travis-CI will execute the build and commit the build to the `deploy` branches, which Azure watches.
## Features:
- Added new Stache commands:
  - `stache release` Mirrors `grunt bump` and pushes to origin
  - `stache commitBuild` Commits build results to deploy branch
- Added Travis-CI config and build scripts
## How to test
### Getting started
1. Ensure that you are the only one reviewing this pull request (some actions below would be destructive to another reviewer).
2. Go to [travis-docs](https://github.com/blackbaud/travis-docs) and delete all branches except `master`.
3. Clone the [travis-docs](https://github.com/blackbaud/travis-docs) repo locally and use it for your tests, since it is already connected to the [Travis-CI dashboard](https://travis-ci.com/blackbaud/travis-docs).
4. Install the Stache CLI `feature-travis-deploy` branch on your local machine, globally (this will replace your existing version of the Stache CLI): 

```
$  npm install -g git+https://github.com/blackbaud/stache-cli.git#feature-travis-deploy
```
## Testing for basic Stache sites:

<img width="196" alt="screen shot 2016-08-26 at 9 47 32 am" src="https://cloud.githubusercontent.com/assets/12497062/18007549/88f07672-6b72-11e6-8aba-88bc9360e424.png">
1. On the [travis-docs](https://github.com/blackbaud/travis-docs) repo, create a new branch named `deploy` from the `master` branch
2. On the Azure Portal, edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy` branch.
3. Clone [travis-docs](https://github.com/blackbaud/travis-docs) locally and make sure the file **stache.deploy.yml** does NOT exist on the root.
4. Next, create a new branch named `feature`.
5. Create an arbitrary commit and push to the `feature` branch.
6. Create a pull request, `master` <-- `feature`, and merge.
7. Travis-CI should build to the `deploy` folder and make a commit to the `deploy` branch.
8. Azure should trigger a redeploy (but **NOT** a build).
9. Make sure that the [Azure site](http://travis.developer.blackbaud.com/) represents your change.
## Testing for Large-scale Stache sites (ie, SKY API):
### Workflow diagram for A-Team/The Flash:

![diagram](https://cloud.githubusercontent.com/assets/12497062/17915501/1064fc0a-6979-11e6-8f84-44733d81a188.jpg)
### Deploying to test site

<img width="276" alt="screen shot 2016-08-26 at 9 47 44 am" src="https://cloud.githubusercontent.com/assets/12497062/18007559/9547df5a-6b72-11e6-850c-0152d917be0d.png">
- On Github, delete all branches except for `master`.
- Clone [travis-docs](https://github.com/blackbaud/travis-docs) locally and open (or create) **stache.deploy.yml**. The file's contents should mirror:

```
env:
  STACHE_DEPLOY_TEST_BRANCH: deploy-test
  STACHE_DEPLOY_PROD_BRANCH: deploy-prod
  STACHE_MASTER_BRANCH: master
  STACHE_DEVELOP_BRANCH: develop
  STACHE_GITHUB_ORG: blackbaud
  STACHE_BUILD_DIRECTORY: deploy
build: <%= stache.config.env.STACHE_BUILD_DIRECTORY %>/
```
- Commit the changes to this file to the `master` branch.
- After the file has been committed to `master`, create the following branches off of the `master` branch:
  - `deploy-test`
  - `deploy-prod`
  - `develop`
- Edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy-test` branch.
- Next, create a new branch from `develop` named `feature`.
- Create an arbitrary commit and push to the `feature` branch.
- Create a pull request, `develop` <-- `feature`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `feature` branch.
### Deploying to production site

<img width="341" alt="screen shot 2016-08-26 at 9 47 50 am" src="https://cloud.githubusercontent.com/assets/12497062/18007563/9bf5a60c-6b72-11e6-8712-838133a4b33c.png">
- (This section depends on the "Deploying to test site" above.)
- Edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy-prod` branch (in a real-life scenario, you would have two separate Azure sites for test and production).
- Next, create a new branch from `develop` named `release`.
- Checkout the `release` branch locally and type `stache release`. Stache CLI will create a commit and push up to `origin release`.
- Create a pull request, `master` <-- `release`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-prod` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure production site](http://travis.developer.blackbaud.com/) represents your change.
- Create a pull request, `develop` <-- `release`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure test site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `release` branch.
### Hotfixes

<img width="337" alt="screen shot 2016-08-26 at 9 47 57 am" src="https://cloud.githubusercontent.com/assets/12497062/18007569/a1441ef4-6b72-11e6-8a8e-ddb0075dbd62.png">
- (This section depends on the "Deploying to production site" above.)
- Create a new branch from `master` named `hotfix`.
- Checkout the `hotfix` branch locally and type `stache release`. Stache CLI will create a commit and push up to `origin hotfix` (in a real-world scenario, an actual change to the source code would come before typing `stache release`).
- Create a pull request, `master` <-- `hotfix`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-prod` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure production site](http://travis.developer.blackbaud.com/) represents your change.
- Create a pull request, `develop` <-- `hotfix`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure test site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `hotfix` branch.
